### PR TITLE
Pypeclub nuke workflow jezschaj

### PR DESCRIPTION
**Pull request by jezscha**
nuke integration contain:
- basic nuke integration (`./setup/nuke` need to be added to **nuke_path**)
- integrated functions in `./avalon/nuke/pipeline` ( ls, publish, containerise, parse_container, update_container, viewer_update_and_undo_stop)
- added some useful functions into `./avalon/nuke/lib` (add_publish_knob, ls_img_sequence, maintained_selection)
- workfiles integration for Nuke
- cgLogging module added to `./vendor` for log support (can be used for maya too)

covering [Nuke workflow #329](https://github.com/getavalon/core/pull/329)

### CODE REVIEW

File: avalon/nuke/pipeline.py [hunk 1]
```diff
@@ -1,43 +1,130 @@
...
+        ("schema", "avalon-core:container-2.0"),
+        ("id", "pyblish.avalon.container"),
+        ("name", str(name)),
+        ("namespace", str(namespace)),
+        ("node_name", str(node_name)),
+        ("representation", str(context["representation"]["_id"])),
+    ]
+
+    try:
+        avalon = node['avalon'].value()
```
[the remark this answers is not in the document]
Comment (jezscha) [reply avalon/nuke/pipeline.py]: just testing if the variable is available

File: avalon/nuke/lib.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
+import re
```
[the remark this answers is not in the document]
Comment (jezscha) [reply avalon/nuke/lib.py]: fixed in new commit

File: avalon/nuke/lib.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import os
+import sys
```
[the remark this answers is not in the document]
Comment (jezscha) [reply avalon/nuke/lib.py]: fixed in next commit

File: avalon/nuke/pipeline.py [hunk 4]
```diff
@@ -48,65 +135,206 @@ def ls():
...
+                project["name"]
+            )
+        )
+        return
+
+    current_width = nuke.root()["format"].value().width()
+    current_height = nuke.root()["format"].value().height()
+    current_pixel_aspect = nuke.root()["format"].value().pixelAspect()
+
+    if width != current_width or height != current_height or current_pixel_aspect != pixel_aspect:
```
[the remark this answers is not in the document]
Comment (jezscha) [reply avalon/nuke/pipeline.py]: fixed in next commit

File: avalon/tools/workfiles/app.py [hunk 5]
```diff
@@ -195,12 +198,16 @@ def setup(self, root):
             "comment": ""
         }
 
-        self.template = "{task[name]}_v{version:0>4}<_{comment}>"
-        templates = self.data["project"]["config"]["template"]
-        if "workfile" in templates:
-            self.template = templates["workfile"]
+        try:
+            self.template = os.environ["AVALON_WORKFILE_TEMPLATE"].replace("\"", "")
```
[the remark this answers is not in the document]
Comment (jezscha) [reply avalon/tools/workfiles/app.py]: fixed in next commit

File: avalon/vendor/cgLogging.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+        msg = self.format(record)
+
+        if record.funcName == "warning":
+            sys.stdout.write(msg+"\n")
+
+        elif record.funcName == "error":
+            sys.stderr.write(msg+"\n")
+
+        elif record.funcName in ["critical", "fatal"]:
+            FBMessageBox(record.funcName, msg, "OK")
```
[the remark this answers is not in the document]
Comment (jezscha) [reply avalon/vendor/cgLogging.py]: comming from Mobu build in functions

File: avalon/vendor/cgLogging.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+# Load Nuke:
+try:
+    import nuke
+    _in_nuke = True
+except Exception:
+    _in_nuke = False
+
+# Load Mobu:
+try:
+    import pyfbsdk
```
[the remark this answers is not in the document]
Comment (jezscha) [reply avalon/vendor/cgLogging.py]: Mobu dependency

File: avalon/vendor/cgLogging.py [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+try:
+    import nuke
+    _in_nuke = True
+except Exception:
+    _in_nuke = False
+
+
+# Load Nuke:
+try:
+    import nuke
```
[the remark this answers is not in the document]
Comment (jezscha) [reply avalon/vendor/cgLogging.py]: testing if in Nuke

File: avalon/nuke/pipeline.py [hunk 9]
```diff
@@ -1,43 +1,130 @@
-"""Pipeline functionality specifically for Nuke
+import os
+import importlib
+from .. import api, io
+import contextlib
+from pyblish import api as pyblish
+from ..vendor import toml
+import nuke
 
-Hi Arvid!
```
Comment (mottosso) [inline avalon/nuke/pipeline.py]: Haha, remnants from the old days. :) Arvid was the Nuke compositor at Mindbender where/when Avalon was born, and this file was made as a template for him to fill in.

File: avalon/nuke/pipeline.py [hunk 10]
```diff
@@ -1,43 +1,130 @@
...
+from ..vendor.cgLogging import getLogger as nLogger
```
Comment (mottosso) [inline avalon/nuke/pipeline.py]: Could you drop a few lines about why you wanted to include `cgLogging` with this PR? What problem(s) were you having that is solved with this library?

Also, let's stick to `snake_case` for our internal members, e.g. `n_logger` or `cg_logger` etc.

File: avalon/tools/workfiles/app.py [hunk 11]
```diff
@@ -195,12 +198,17 @@ def setup(self, root):
             "comment": ""
         }
 
-        self.template = "{task[name]}_v{version:0>4}<_{comment}>"
-        templates = self.data["project"]["config"]["template"]
-        if "workfile" in templates:
-            self.template = templates["workfile"]
+        try:
+            self.template = os.environ["AVALON_WORKFILE_TEMPLATE"]\
```
Comment (mottosso) [inline avalon/tools/workfiles/app.py]: Where does `AVALON_WORKFILE_TEMPLATE` come from? This normally resides in the database.

File: avalon/vendor/toml.py [hunk 12]
```diff
@@ -4,9 +4,11 @@
 import datetime
 import io
 
+
```
Comment (mottosso) [inline avalon/vendor/toml.py]: Best not to edit vendor files, we can rely on them being tested in their own project and CI pipeline. Any edits we make would be overwritten whenever we update the library anyway.

Was it throwing errors on Travis? They should be excluded from our tests.

**Comment by tokejepsen** [comment]
Could I ask why a new PR and not work from the existing PR?

Also could you give a workflow breakdown of the integration, like what are containers etc?

File: avalon/nuke/pipeline.py (hunk 9 above)
Comment (jezscha) [reply avalon/nuke/pipeline.py]: :) will remove the trace.

File: avalon/pipeline.py [hunk 13]
```diff
@@ -156,17 +156,17 @@ class Loader(list):
...
-        template = context["project"]["config"]["template"]["publish"]
 
-        data = {
-            key: value["name"]
-            for key, value in context.items()
-        }
 
-        data["root"] = registered_root()
-        data["silo"] = context["asset"]["silo"]
+        try:
```
[the remark this answers is not in the document]
Comment (mottosso) [reply avalon/pipeline.py]: https://www.youtube.com/watch?v=sqcLjcSloXs

**Comment by jezscha** [comment]
> Could I ask why a new PR and not work from the existing PR?

@tokejepsen yep that is good suggestion. No offence :)

**Comment by tokejepsen** [comment]
> No offence

None taken :) Just wondered whether there was something different from the other PR, so I can potentially close the other PR for the same functionality.

**Pull request closed without merging**